### PR TITLE
Wip user specified flushing

### PR DESCRIPTION
**Pull request by dragonylffly**
This patch implements user specified cache flushing for cache tiering. In 
many use cases, the user traffic is high in only some fixed time, fox example,
work hour. In those cases, it is desirable that the flushing job could be scheduled 
in the idle time, and do as much as possible, to reduce the bandwidth contention
with user traffic.

Signed-off-by: Mingxin Liu <mingxinliu@ubuntukylin.com>
Reviewed-by: Li Wang <liwang@ubuntukylin.com>
Suggested-by: Xinze Chi <xmdxcxz@gmail.com>

### CODE REVIEW

**Comment by shishirng** [comment]
Looks good to me.

File: src/osd/OSD.h [hunk 1]
```diff
@@ -658,12 +658,18 @@ class OSDService {
...
-  Mutex agent_timer_lock;
-  SafeTimer agent_timer;
-
   void agent_entry();
   void agent_stop();
 
+  bool agent_in_schedule_time;
+  Mutex agent_schedule_lock;
+  int agent_schedule_flush_version;
+  Mutex agent_timer_lock;
```
Comment (liewegas) [inline src/osd/OSD.h]: Can't we reuse agent_timer_lock instead of adding an agent_schedule_lock?

**Comment by liewegas** [comment]
I'm not a huge fan of the way this is structured.  Partly it's because it's a bit kludgey currently, but it would really nice if we could unify the active/inactive state of the agent with the current idle mode (where it'll pause for a bit when it doesn't find any work to do).  Otherwise, the combination of that, the restart flag to agent_choose_mode(), and this new time-based scheduling make it really hard to understand the state of the agent.  Any ideas on how we can make it easier to follow?

**Comment by dragonylffly** [comment]
@liewegas  I think this is potentially very useful since the cache pool is typically large enough to hold the dirty data generated within one day, then this feature could leave the user traffic not to be contended. Regarding the agent state, it seems clear. The agent is idle in two situations, one is either the pg without enough dirty objects, or not in the scheduled flushing time if scheduled flushing enabled; The other is the statistics are not trustable, in that case, it dequeues the pg from flushing queue, and schedules a delayed recall to agent_choose_mode() with the restart parameter being true to avoid dequeue the already dequeued pg. Other states, FLUSH_MODE_LOW, FLUSH_MODE_HIGH are clear. Maybe we add FLUSH_MODE_ALL to indicate the agent will flush all the dirty objects in the sceduled time, what is your opinion?

**Comment by dachary** [comment]
@dragonylffly this needs rebasing

**Comment by dachary** [comment]
@dragonylffly feel free to re-open when you have time to rebase

**Pull request closed without merging**